libs/sdk-ui/src/base/react/placeholders/PlaceholdersContext.tsx

```diff
@@ -0,0 +1,441 @@
+// (C) 2019 GoodData Corporation
+import { v4 as uuidv4 } from "uuid";
+import React, { createContext, useCallback, useContext, useState } from "react";
+import noop from "lodash/noop";
+import { ValueOrUpdateCallback } from "@gooddata/sdk-backend-base";
+import {
+    IAttributeGroupPlaceholder,
+    IAttributePlaceholder,
+    IComputedPlaceholder,
+    IFilterGroupPlaceholder,
+    IFilterPlaceholder,
+    IMeasureGroupPlaceholder,
+    IPlaceholderWithHook,
+    isPlaceholder,
+    PlaceholdersReturnTypes,
+    UndefinedPlaceholderHandling,
+    IMeasurePlaceholder,
+    IPlaceholder,
+    PlaceholderReturnType,
+    ValueWithPlaceholdersReturnType,
+    ValueWithPlaceholders,
+    isGroupPlaceholder,
+    ValuesWithPlaceholdersReturnTypes,
+    PlaceholderComputation,
+} from "./interfaces";
+import isArray from "lodash/isArray";
+import { IAttribute, IAttributeOrMeasure, IFilter } from "@gooddata/sdk-model";
+import flatMap from "lodash/flatMap";
+import { UnexpectedSdkError } from "../../errors/GoodDataSdkError";
+
+export type PlaceholdersState = Record<string, IPlaceholder>;
+
+/**
+ * @internal
+ */
+interface IPlaceholdersContextState {
+    placeholders: PlaceholdersState;
+    updatePlaceholders: (callback: (state: PlaceholdersState) => PlaceholdersState) => void;
+    undefinedPlaceholderHandling: UndefinedPlaceholderHandling;
+}
+
+/**
+ * @internal
+ */
+const PlaceholdersContext = createContext<IPlaceholdersContextState>({
+    undefinedPlaceholderHandling: "none",
+    placeholders: {},
+    updatePlaceholders: noop,
+});
+PlaceholdersContext.displayName = "PlaceholdersContext";
+
+/**
+ * @public
+ */
+export interface IPlaceholdersProviderProps {
+    children: React.ReactNode;
+    undefinedPlaceholderHandling?: UndefinedPlaceholderHandling;
+    debug?: boolean;
+}
+
+/**
+ *
+ * @public
+ */
+export function PlaceholdersProvider(props: IPlaceholdersProviderProps) {
+    const { undefinedPlaceholderHandling = "none", debug, children } = props;
+    const [placeholders, updatePlaceholders] = useState<PlaceholdersState>({});
+    if (debug) {
+        console.debug("[DEBUG]: Current placeholders state:", placeholders);
```

Comment:
Would https://reactjs.org/docs/hooks-reference.html#usedebugvalue be useful here? Maybe even unconditionally (that could remove the debug prop).